### PR TITLE
Resolve Embedding and Setting deserialization bug

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12704?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12704/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12704
```

</p>
</details>

### What changes are proposed in this pull request?

While testing the flavor with Databricks LLMs/embeddings, found a few issues.
1. Embedding model cannot be instantiated due to an error `TypeError: pydantic.main.BaseModel.__init__() got multiple values for keyword argument 'model_name`. This is an inconsistency in LlamaIndex side where `BaseEmbedding` class has `model_name` as constructor argument while many implementation like `DatabricksEmbeddings` has `model`. It's not feasible to solve this for all community classes in llama index so we have to handle this in our logic.
2. Setting is not deserialized when loading pyfunc. It is deserialized when loading via `mlflow.llama_index.load_model`, but not via `mlflow.pyfunc.load_model`. This is because the `deserialize_settings()` is not called within `_load_pyfunc()`.

This PR fixes the both issues.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Tested index loading with `Databricks` LLM and `DatabricksEmbedding`.

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/llama_index/__init__.py
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:341]: This solves the issue #1. We should load the index via `load_model()` that includes `deserialize_setting` call, rather than `_load_index()`.

File: mlflow/llama_index/serialize_objects.py
Comment (B-Step62) [inline mlflow/llama_index/serialize_objects.py:119]: `isinstance` only works between isntance and class and we must use `issubsclass` here.

File: mlflow/llama_index/serialize_objects.py
Comment (B-Step62) [inline mlflow/llama_index/serialize_objects.py:137]: this solves the issue 2

File: mlflow/llama_index/serialize_objects.py
Comment (B-Step62) [inline mlflow/llama_index/serialize_objects.py:17]: @BenWilson2 @michael-berk This mapping has never been used for deserialization due to [the issubclass bug](https://github.com/mlflow/mlflow/pull/12704/files#r1682197429), but we haven't face a problem as LlamaIndex defines every components as pydantic model. Taking that standardization into account, I'm wondering if we need this extra complexity in our code base. It's also hard to test with high coverage, like we haven't caught the issubclass bug until now.. Curious to hear you guys' thought.

**Review by michael-berk — APPROVED** [review]

File: mlflow/llama_index/serialize_objects.py
Comment (michael-berk) [reply mlflow/llama_index/serialize_objects.py:17]: The original implementation simply looked for `hasattr(obj, "to_dict")`/`hasattr(obj, "from_dict")` without type validation, but we refactored because
1. We could not guarantee that those methods were the globally correct serde methods
2. Checking types was consistent with other mlflow flavor implementations (langchain)

Based on those two points, I think the mapping still has validity (with the `subclass` bug fix). But, I definitely defer to you two

Comment (B-Step62) [reply mlflow/llama_index/serialize_objects.py:17]: Yes, the first one was my original understanding about the reasoning. However, it doesn't seem to be practical for us to constantly check any new object types in their repo and maintain the curated list to be accurate, meaning that even having a fixed list does not help the reliability (it's still human in the end either they or we maintain the logic).

> Checking types was consistent with other mlflow flavor implementations (langchain)

My perspective on this is indeed opposite. Managing the serialization logic in LangChain flavor has been a big pain for us. Imo we don't necessarily need consistency with what we did a year ago as a first trial, rather implement a better way from the learning.

Any thought? @BenWilson2 ?

**Comment by B-Step62** [comment]
Merging to unblock bug bash ([the open comment](https://github.com/mlflow/mlflow/pull/12704#discussion_r1682254152) is rather a follow-up discussion)

File: mlflow/llama_index/serialize_objects.py
Comment (BenWilson2) [reply mlflow/llama_index/serialize_objects.py:17]: Let's leverage the built-in functionality within pydantic and let LlamaIndex handle this. 
If there is an issue with calling a serialization operation on something that they've built, our standard process for other libraries is to file an issue in their repo to give them a heads up that they just broke something. We don't need to nanny-check their implementations within our integration. 
Let's strip this out.

**Pull request merged**
